### DISCUSSION TITLE
Intermittent SSL Cert error in "Microsoft SQL Server Management Studio" while using the driver in combination with a linked server

### DISCUSSION THREAD

**Discussion by AparnaSuriaraj18** [openssl/openssl#23355] [Q&A]
in "Microsoft SQL Server Management Studio" while running the query using the driver in
combination with a MS SQL Server linked server.
The error returned is:
OLE DB provider "STREAM" for linked server "(null)" returned message "[Simba][DriverSupport] (1210) Unable to load Client certificate or Client private key set via the connection attributes ClientCert or ClientPrivateKey. Please ensure the file exists, has read permission, and has the correct PEM format.".
We get the error from the curl side wherein the function curl_easy_perform() returns the curl error code "CURLE_SSL_CERTPROBLEM (58)" when this error happens we could see that the correct
cert file path is getting picked but issue might be occurring when the contents of the cert file is read. There are multiple SSL handshakes that takes place and in the initial handshake it is able to read the
ssl file but it fails later with the error
OpenSSL error error:80000005:system library::Input/output error, (no key found, wrong pass phrase, or wrong file format?)
from the openssl.c file present in the curl thirdparty library.
This is an intermittent issue, which is observed only in "Microsoft SQL Server Management Studio" but
the ODBC driver works fine outside of a linked server - e.g. the same queries are fine if run through SSIS/PowerShell,odbctest tool etc.

Any pointer that can help us in understanding this inconsistent behavior and resolving this intermittent error that we are getting.

Thanks in advance!

**Comment by tom-cosgrove-arm** [comment]
I suspect this would be better as a discussion...

**Reply by t8m** [reply]
Yes, questions must be posted as discussions in the Q&A section.

**Comment by mattcaswell** [comment]
Which version of OpenSSL are you using?

**Reply by AparnaSuriaraj18** [reply]
We are using openssl-3.0

**Comment by mattcaswell** [comment]
This error message is the OpenSSL specific bit:

````
OpenSSL error error:80000005:system library::Input/output error
````

The `8` in that simply tells you that this is a "system" error, i.e. OpenSSL has made some system call and got an error back. The 5 tells you what that error is:

https://learn.microsoft.com/en-us/cpp/c-runtime-library/errno-constants?view=msvc-170

````
EIO	I/O error.
````

The rest of that error line comes from curl:

https://github.com/curl/curl/blob/f81a335e85ba625c92da5738cb7b693cd71662ad/lib/vtls/openssl.c#L1297-L1334

Basically OpenSSL has been asked to open a file to read a certificate from it, but we've got an IO error back when attempting to do so.

The reasons for why that occur are almost certainly "higher" up in the software stack, i.e. probably not an issue in OpenSSL itself.

**Reply by AparnaSuriaraj18** [reply]
This error message is coming from certstuff() in openssl.c file from curl library side. But the condition that is being checked before returning this error is **if(cert_use_result != 1)** 
The value of cert_use_result is set to 0 in **SSL_CTX_use_certificate_chain_file** which is from the openssl side.
